### PR TITLE
Move pattern information to separate YAML files

### PR DESCRIPTION
**Pull request by yousefmoazzam**
This PR is intended to address #61.

The main change is to replace the `@pattern` decorator (which was used on method functions to associate patterns to the function objects), with storing pattern information in separate YAML files and fetching this information in the runner.

### CODE REVIEW

**Review by dkazanc — APPROVED** [review]

**Comment by namannimmo10** [comment]
Please make sure to run the tests before pushing the changes to the `main` branch -- tests fail with this branch.

```console
$ pytest
============================================ test session starts ============================================
platform linux -- Python 3.9.13, pytest-7.2.1, pluggy-1.0.0 -- /dls/science/users/qfc83269/condaenvs/httomo/bin/python3.9
cachedir: .pytest_cache
rootdir: /home/qfc83269/Applications/oss/httomo, configfile: pyproject.toml, testpaths: tests
plugins: anyio-3.5.0, xdist-3.1.0, cov-4.0.0
collected 11 items                                                                                          

tests/test_cli.py::test_cli_version_shows_version PASSED                                              [  9%]
tests/test_cli.py::test_cli_help_shows_help PASSED                                                    [ 18%]
tests/test_cli.py::test_cli_noargs_raises_error PASSED                                                [ 27%]
tests/test_loader.py::test_tomo_standard_loaded PASSED                                                [ 36%]
tests/test_loader.py::test_tomo_standard_testing_pipeline_loaded FAILED                               [ 45%]
tests/test_loader.py::test_tomo_standard_testing_pipeline_loaded_with_save_all FAILED                 [ 54%]
tests/test_loader.py::test_k11_diad_loaded PASSED                                                     [ 63%]
tests/test_loader.py::test_diad_testing_pipeline_loaded FAILED                                        [ 72%]
tests/test_pipeline.py::test_tomo_standard_testing_pipeline_output FAILED                             [ 81%]
tests/test_pipeline.py::test_tomo_standard_testing_pipeline_output_with_save_all FAILED               [ 90%]
tests/test_pipeline.py::test_diad_testing_pipeline_output FAILED                                      [100%]

================================================= FAILURES ==================================================
________________________________ test_tomo_standard_testing_pipeline_loaded _________________________________
Traceback (most recent call last):
  File "/home/qfc83269/Applications/oss/httomo/tests/test_loader.py", line 30, in test_tomo_standard_testing_pipeline_loaded
    output = subprocess.check_output(cmd).decode().strip()
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/dls/science/users/qfc83269/condaenvs/httomo/bin/python3.9', '-m', 'httomo', '--ncore', '2', 'tests/test_data/tomo_standard.nxs', 'temp.yaml', 'output_dir/', 'task_runner']' returned non-zero exit status 1.
------------------------------------------- Captured stderr call --------------------------------------------
Traceback (most recent call last):
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/qfc83269/Applications/oss/httomo/httomo/__main__.py", line 4, in <module>
    main()
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/decorators.py", line 38, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/qfc83269/Applications/oss/httomo/httomo/cli.py", line 85, in task_runner
    run_tasks(
  File "/home/qfc83269/Applications/oss/httomo/httomo/task_runner.py", line 102, in run_tasks
    _assign_pattern_to_method(func, module_path, params['method_name'])
  File "/home/qfc83269/Applications/oss/httomo/httomo/task_runner.py", line 760, in _assign_pattern_to_method
    pattern_str = get_method_info(module_path, method_name, 'pattern')
  File "/home/qfc83269/Applications/oss/httomo/httomo/methods_database/query.py", line 43, in get_method_info
    info = info[key]
KeyError: 'stripe'
_________________________ test_tomo_standard_testing_pipeline_loaded_with_save_all __________________________
Traceback (most recent call last):
  File "/home/qfc83269/Applications/oss/httomo/tests/test_loader.py", line 51, in test_tomo_standard_testing_pipeline_loaded_with_save_all
    output = subprocess.check_output(cmd).decode().strip()
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/dls/science/users/qfc83269/condaenvs/httomo/bin/python3.9', '-m', 'httomo', '--save_all', '--ncore', '2', 'tests/test_data/tomo_standard.nxs', 'temp.yaml', 'output_dir/', 'task_runner']' returned non-zero exit status 1.
------------------------------------------- Captured stderr call --------------------------------------------
Traceback (most recent call last):
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/qfc83269/Applications/oss/httomo/httomo/__main__.py", line 4, in <module>
    main()
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/dls/science/users/qfc83269/condaenvs/httomo/lib/python3.9/site-packages/click/decorators.py", line 38, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/qfc83269/Applications/oss/httomo/httomo/cli.py", line 85, in task_runner
    run_tasks(
  File "/home/qfc83269/Applications/oss/httomo/httomo/task_runner.py", line 102, in run_tasks
    _assign_pattern_to_method(func, module_path, params['method_name'])
  File "/home/qfc83269/Applications/oss/httomo/httomo/task_runner.py", line 760, in _assign_pattern_to_method
    pattern_str = get_method_info(module_path, method_name, 'pattern')
  File "/home/qfc83269/Applications/oss/httomo/httomo/methods_database/query.py", line 43, in get_method_info
    info = info[key]
KeyError: 'stripe'
```

File: httomo/methods_database/query.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    yaml_info_path = Path(YAML_DIR, f"{package_name}.yaml")
+
+    if not yaml_info_path.exists():
+        err_str = f"The YAML file {yaml_info_path} doesn't exist."
+        raise ValueError(err_str)
+
+    with open(yaml_info_path, 'r') as f:
+        info = yaml.safe_load(f)
+        for key in split_method_path[1:]:
+            info = info[key]
```
Comment (team-gpu) [inline httomo/methods_database/query.py:43]: If the requested method is not in the yaml database file, this fails with a cryptic `KeyError`. I think this should be explicitly checked and a more user-friendly error generated.

File: httomo/methods_database/packages/tomopy.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+prep:
+  phase:
+    retrieve_phase:
+      pattern: projection
+  normalize:
+    minus_log:
+      pattern: projection
+    normalize:
+      pattern: projection
```
Comment (team-gpu) [inline httomo/methods_database/packages/tomopy.yaml:9]: This is needed to run the sample that Daniil provided a few days ago...

```yaml
  stripe:
    remove_all_stripe:
      pattern: projection
```

File: httomo/methods_database/query.py (hunk 1 above)
Comment (yousefmoazzam) [reply httomo/methods_database/query.py:43]: Yep, I completely agree with this point. Raising an error would help methods developers who have maybe missed adding in an entry when they have added in a new method, and would help users if they have accidentally made a mistake in the YAML (though ideally, some kind of validation of the YAML file being given to httomo will be done in the future to handle this user-error case).

**Pull request merged**